trie/triedb/pathdb/nodebufferlist.go

```diff
@@ -0,0 +1,864 @@
+		for path, n := range subset {
+			size += uint64(len(n.Blob) + len(path))
+		}
+	}
+	base := newMultiDifflayer(limit, size, common.Hash{}, nodes, layers)
+	ele := newMultiDifflayer(limit, 0, common.Hash{}, make(map[common.Hash]map[string]*trienode.Node), 0)
+	nf := &nodebufferlist{
+		db:        db,
+		wpBlocks:  wpBlocks,
+		rsevMdNum: rsevMdNum,
+		dlInMd:    dlInMd,
+		limit:     limit,
+		base:      base,
+		head:      ele,
+		tail:      ele,
+		count:     1,
+		persistID: rawdb.ReadPersistentStateID(db),
+		stopCh:    make(chan struct{}),
+	}
+	go nf.loop()
+
+	log.Info("new node buffer list", "proposed block interval", nf.wpBlocks,
+		"reserve multi difflayers", nf.rsevMdNum, "difflayers in multidifflayer", nf.dlInMd,
+		"limit", common.StorageSize(limit), "layers", layers, "persist id", nf.persistID)
+	return nf
+}
+
+// node retrieves the trie node with given node info.
+func (nf *nodebufferlist) node(owner common.Hash, path []byte, hash common.Hash) (node *trienode.Node, err error) {
+	nf.mux.RLock()
+	find := func(nc *multiDifflayer) bool {
+		subset, ok := nc.nodes[owner]
+		if !ok {
+			return true
+		}
+		n, ok := subset[string(path)]
+		if !ok {
+			return true
+		}
+		if n.Hash != hash {
+			log.Error("Unexpected trie node in node buffer list", "owner", owner, "path", path, "expect", hash, "got", n.Hash)
+			err = newUnexpectedNodeError("dirty", hash, n.Hash, owner, path, n.Blob)
+			return false
+		}
+		node = n
+		return false
+	}
+	nf.traverse(find)
+	if err != nil {
+		nf.mux.RUnlock()
+		return nil, err
+	}
+	if node != nil {
+		nf.mux.RUnlock()
+		return node, nil
+	}
+	nf.mux.RUnlock()
+
+	nf.baseMux.RLock()
+	node, err = nf.base.node(owner, path, hash)
+	nf.baseMux.RUnlock()
+	return node, err
+}
+
+// commit merges the dirty nodes into the trienodebuffer. This operation won't take
+// the ownership of the nodes map which belongs to the bottom-most diff layer.
+// It will just hold the node references from the given map which are safe to
+// copy.
+func (nf *nodebufferlist) commit(root common.Hash, id uint64, block uint64, nodes map[common.Hash]map[string]*trienode.Node) trienodebuffer {
+	nf.mux.Lock()
+	defer nf.mux.Unlock()
+
+	if nf.head == nil {
+		nf.head = newMultiDifflayer(nf.limit, 0, common.Hash{}, make(map[common.Hash]map[string]*trienode.Node), 0)
+		nf.tail = nf.head
+	}
+	oldSize := nf.head.size
+	err := nf.head.commit(root, id, block, 1, nodes)
+	if err != nil {
+		log.Crit("failed to commit nodes to node buffer list", "error", err)
+	}
+
+	nf.stateId = id
+	nf.block = block
+	nf.size = nf.size + nf.head.size - oldSize
+	nf.layers++
+
+	nodeBufferListSizeGauge.Update(int64(nf.size))
+	nodeBufferListLayerGauge.Update(int64(nf.layers))
+	nodeBufferListLastStateIdGauge.Update(int64(nf.stateId))
+	nodeBufferListLastBlockGauge.Update(int64(nf.block))
+
+	if block != 0 && block%nf.dlInMd == 0 {
+		nc := newMultiDifflayer(nf.limit, 0, common.Hash{}, make(map[common.Hash]map[string]*trienode.Node), 0)
```

Comment:
I suspect that the conditions on lines 187 and 167 may sometimes be satisfied at the same time, resulting in the creation of multiple MultiDifflayers. Is this expected?